### DISCUSSION TITLE
How should I handle dependencies between loaders?

### DISCUSSION THREAD

**Discussion by adrianmroz-allegro** [remix-run/remix#2042] [Q&A] [answered]
Hi!

I can't share specifics of my application, so let's pretend that I'm writing HR app for international company. I have this view with three sections:

* Three inputs for Country (multi select), start date and end date. Let's call them "general filters".
* Multiple filters that can select workers, like sex, nationality, seniority etc. Let's call them "people filters".
* Table that displays information about workers that fulfilled filter criteria.

Filters can't really be serialised to route params, they must be represented as query params (I think, correct me if there is a way to do it).

So I have this one route with a bunch of query params. If user changes Country select box, we need to fetch "people filters" (their value depends on "general filters" - for example we show only nationalities of existing people in selected country) and then fetch people data. Simple. 

Problem is if I change something in "people filters" - I don't want to refetch all "general filters" values. Often users would click around many "people filters" controls and emit a lot of route changes. Changing "people filters" can't affect "general filters" - they are always the same. Can I do this at remix level? 

Thanks for help!

**Comment by kiliman** [comment, marked as answer]
In Remix each route can have a loader. The loader will *not* be called again until the user navigates to another route that doesn't have a common ancestor with the current route.

That is if you navigate from `/parent/child-a` to `/parent/child-b` then only the loader for `child-b` will be called since they share `root` and `parent` routes.

So in your scenario, you should create a common parent route that returns the filter data. If you don't want it to affect your URL, you can use a *pathless* route. 

As you update the search params of `__filter/table` route, it will *not* reload the data from the parent loaders since the parent route hasn't changed.

Something like:
```ts
// __filter.tsx  <- loader returns filter lists (`__` prefix means pathless, doesn't affect URL)
export const function loader() {
  return json({ general: { countries: [...], people: { workers: [...], sex: [...], ... }})
}

// __filter/table.tsx  <- loader uses filter values from URLSearchParams to filter and return data
export const function loader({request) {
  const url = new URL(request.url)
  const countries = url.searchParams.getAll('countries')
  const ...
  const filteredData = getData(countries, workers, ...)
  return json(filteredData)
}

export default function Table() {
  const filteredData = useLoaderData()
  const matches = useMatches()  // use matches lets us get any parent data
  const filterList = matches.find(route => route.id === 'routes/__filter').data
  ...
```

**Reply by adrianmroz-allegro** [reply]
Hi! Thanks for idea with pathless routes, that's genius! I will try implement this in my case.
